src/System.Private.CoreLib/src/System/ValueType.cs

```diff
@@ -36,4 +42,161 @@ public override int GetHashCode()
+
+            if (numFields == UseFastHelper)
+            {
+                hashCode ^= FastGetValueTypeHashCodeHelper(this.EETypePtr, ref this.GetRawData());
+            }
+            else
+            {
+                hashCode ^= RegularGetValueTypeHashCode(this.EETypePtr, ref this.GetRawData(), numFields);
+            }
+
+            return hashCode;
+        }
+
+        private static int FastGetValueTypeHashCodeHelper(EETypePtr type, ref byte data)
+        {
+            // Sanity check - if there are GC references, we should not be hashing bytes
+            Debug.Assert(!type.HasPointers);
+
+            int size = (int)type.ValueTypeSize;
+            int hashCode = 0;
+
+            for (int i = 0; i < size / 4; i++)
+            {
+                hashCode ^= Unsafe.As<byte, int>(ref Unsafe.Add(ref data, i * 4));
+            }
+
+            return hashCode;
+        }
+
+        private int RegularGetValueTypeHashCode(EETypePtr type, ref byte data, int numFields)
+        {
+            int hashCode = 0;
+
+            // We only take the hashcode for the first non-null field. That's what the CLR does.
+            for (int i = 0; i < numFields; i++)
+            {
+                int fieldOffset = __GetFieldHelper(i, out EETypePtr fieldType);
+                ref byte fieldData = ref Unsafe.Add(ref data, fieldOffset);
+
+                Debug.Assert(!fieldType.IsPointer);
+
+                if (fieldType.CorElementType == RuntimeImports.RhCorElementType.ELEMENT_TYPE_R4)
+                {
+                    hashCode = Unsafe.Read<float>(ref fieldData).GetHashCode();
+                }
+                else if (fieldType.CorElementType == RuntimeImports.RhCorElementType.ELEMENT_TYPE_R8)
+                {
+                    hashCode = Unsafe.Read<double>(ref fieldData).GetHashCode();
+                }
+                else if (fieldType.IsPrimitive)
+                {
+                    hashCode = FastGetValueTypeHashCodeHelper(fieldType, ref fieldData);
+                }
+                else if (fieldType.IsValueType)
+                {
+                    // We have no option but to box and call regular GetHashCode since this value type could have
+                    // GC pointers (we could find out if we want though), or fields of type Double/Single (we can't
+                    // really find out). Double/Single have weird requirements around -0.0 and +0.0.
+                    // If this boxing becomes a problem, we could build a piece of infrastructure that determines the slot
+                    // of __GetFieldHelper, decodes the unboxing stub pointed to by the slot to the real target
+                    // (we already have that part), and calls the entrypoint that expects a byref `this`, and use the
+                    // data to decide between calling fast or regular hashcode helper.
+                    object fieldValue = RuntimeImports.RhBox(fieldType, ref fieldData);
```

Comment:
Same boxing issue

Reply:
The current CoreCLR implementation won't actually box and call GetHashCode for valuetype fields. It will always take xor paths for them (ignoring any GetHashCode overrides that the valuetypes may have).

Reply:
> It will always take xor paths for them

It won't take the XOR path if the nested struct contains GC pointers or floating-point fields. Without building the extra feature the comment is talking about, we can't avoid the box. I'll do some factoring to avoid calling the actual `GetHashCode` and simulate what CLR would do.